File: protovalidate/grpc-python/start/invoice_server.py
# Copyright 2025 Buf Technologies, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import grpc
import invoice.v1.invoice_service_pb2 as invoice_service_pb2
import invoice.v1.invoice_service_pb2_grpc as invoice_service_pb2_grpc
import logging
import sys

from concurrent import futures
from grpc_reflection.v1alpha import reflection
from invoice.v1.invoice_service import InvoiceService

# Configure a logger that won't show Protovalidate logging.
logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)
handler = logging.StreamHandler(sys.stdout)
handler.setFormatter(
    logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
)
logger.addHandler(handler)

# Server configuration
port = "50051"


def serve():
    server = grpc.server(
        futures.ThreadPoolExecutor(max_workers=10),
    )
    invoice_service_pb2_grpc.add_InvoiceServiceServicer_to_server(
        InvoiceService(), server
    )
    SERVICE_NAMES = (
        invoice_service_pb2.DESCRIPTOR.services_by_name["InvoiceService"].full_name,
        reflection.SERVICE_NAME,
    )
    reflection.enable_server_reflection(SERVICE_NAMES, server)

    server.add_insecure_port("[::]:" + port)
    logger.info("Invoice server started on port " + port)
    server.start()
    server.wait_for_termination()


if __name__ == "__main__":
    serve()

File: protovalidate/grpc-java/finish/src/main/java/invoice/v1/InvoiceServer.java
// Copyright 2025 Buf Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package invoice.v1;

import buf.build.example.protovalidate.ValidationInterceptor;
import build.buf.protovalidate.Validator;
import io.grpc.*;
import io.grpc.protobuf.services.ProtoReflectionServiceV1;
import java.io.IOException;
import java.util.concurrent.TimeUnit;
import java.util.logging.Logger;

/**
 * InvoiceServer provides a class that manages a gRPC server and a main() entry point for its
 * startup.
 */
public class InvoiceServer {

    /** logger is a Logger for use by this class. */
    private static final Logger log = Logger.getLogger(InvoiceServer.class.getName());

    /** gRPC server is an io.grpc.Server. */
    private final Server gRPCServer;

    /** port is the port on which the server should listen. */
    private final int port;

    /**
     * Creates an InvoiceServer
     *
     * @param port - port to listen on
     * @param serverCredentials - ServerCredentials implementation
     */
    public InvoiceServer(int port, ServerCredentials serverCredentials, ServerServiceDefinition service) {
        this.port = port;

        gRPCServer = Grpc.newServerBuilderForPort(port, serverCredentials)
                .addService(service)
                .addService(ProtoReflectionServiceV1.newInstance())
                .build();
    }

    protected void run() throws IOException {
        gRPCServer.start();

        log.info("Server started on port " + port);

        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            try {
                InvoiceServer.this.shutdown();
                System.out.println("Server shut down");
            } catch (Exception e) {
                System.err.println("Error stopping server");
                e.printStackTrace(System.err);
            }
        }));
    }

    protected void shutdown() throws InterruptedException {
        if (!gRPCServer.isTerminated()) {
            gRPCServer.shutdown().awaitTermination(1L, TimeUnit.MINUTES);
        }
    }

    private void awaitTermination() throws InterruptedException {
        if (!gRPCServer.isTerminated()) {
            gRPCServer.awaitTermination();
        }
    }

    public static void main(String[] args) throws IOException, InterruptedException {
        final BindableService service = new InvoiceService();
        final ValidationInterceptor interceptor = new ValidationInterceptor(new Validator());
        final ServerServiceDefinition serviceDefinition = ServerInterceptors.intercept(service, interceptor);

        final InvoiceServer invoiceServer =
                new InvoiceServer(50051, InsecureServerCredentials.create(), serviceDefinition);

        invoiceServer.run();
        invoiceServer.awaitTermination();
    }
}
